### PR TITLE
[F#] Improvements in asm resolution and tests

### PR DESCRIPTION
**Pull request by mhutch**
* Simplify F# assembly resolution to lean more heavily on MSBuild, improving consistency between the in-process compiler and the build.
* Try to pass state as args instead of pulling from global state are various points in call hierarchy. Fixes possible inconsistent state in a few places.
* Eliminate wasteful duplicated assembly resolution calls and blocking on async code.
* Have tests use the shared helpers a bit more.
* Fix support for `PackageReference`

### CODE REVIEW

**Comment by mhutch** [comment]
(cherry-picked from https://github.com/mono/monodevelop/pull/5363)

**Review by nosami — APPROVED** [review]

File: main/external/fsharpbinding/MonoDevelop.FSharp.Shared/Extensions.fs [hunk 1]
```diff
@@ -9,6 +10,11 @@ open ExtCore
 module Seq =
     let tryHead items =
         if Seq.isEmpty items then None else Some (Seq.head items)
+
+    let ofType<'T> (items:IEnumerable) : seq<'T> =
+        items
+        |> Seq.cast<obj>
```
[the remark this answers is not in the document]
Comment (mhutch) [reply main/external/fsharpbinding/MonoDevelop.FSharp.Shared/Extensions.fs]: If I do that I don't seem to be able to apply it to .NET collections

File: main/external/fsharpbinding/MonoDevelop.FSharpBinding/Services/LanguageService.fs [hunk 2]
```diff
@@ -272,10 +272,15 @@ type LanguageService(dirtyNotify, _extraProjectInfo) as x =
...
 
-    let optionsForDependentProject p =
+    let optionsForDependentProject projectFile =
+        let project = x.GetProjectFromFileName projectFile
         async {
-            let! assemblies = x.GetReferencedAssembliesAsync p
-            return x.GetProjectCheckerOptions(p, [], assemblies)
+            let! assemblies = async {
+                match project with
+                | Some (proj:DotNetProject) -> return! proj.GetReferences(CompilerArguments.getConfig()) |> Async.AwaitTask
```
[the remark this answers is not in the document]
Comment (mhutch) [reply main/external/fsharpbinding/MonoDevelop.FSharpBinding/Services/LanguageService.fs]: Yes. Not sure why. Without it `proj` shows correct type (in tooltip) but the GetReferences call doesn't resolve.

**Review by slluis — APPROVED** [review]

File: main/external/fsharpbinding/MonoDevelop.FSharp.Tests/MonoDevelop.FSharp.Tests.fsproj [hunk 3]
```diff
@@ -166,6 +158,7 @@
     <ProjectReference Include="..\..\..\tests\UnitTests\UnitTests.csproj">
       <Project>{1497D0A8-AFF1-4938-BC22-BE79B358BA5B}</Project>
       <Name>UnitTests</Name>
+      <Private>False</Private>
```
Comment (Therzok) [inline main/external/fsharpbinding/MonoDevelop.FSharp.Tests/MonoDevelop.FSharp.Tests.fsproj]: This should've stayed private.

Comment (mhutch) [reply main/external/fsharpbinding/MonoDevelop.FSharp.Tests/MonoDevelop.FSharp.Tests.fsproj]: The output dir is the same.

**Pull request merged**